grakn-client/src/main/java/ai/grakn/client/BatchMutatorClient.java

```diff
@@ -275,41 +286,36 @@ void sendQueriesToLoader(Collection<Query> queries){
         try {
             taskId = sendQueryRetry.call(callable);
         } catch (Exception e) {
-            LOG.error("Error while executing queries:\n{}", queries);
+            printError("Error while executing queries:\n{" + queries + "}");
             throw new RuntimeException(e);
         }
 
         CompletableFuture<Json> status = makeTaskCompletionFuture(taskId);
 
         // Add this status to the set of completable futures
-        // TODO: use an async client
-        futures.put(status.hashCode(), status);
 
-        status
-        // Unblock and log errors when task completes
-        .handle((result, error) -> {
-            unblock(status);
+        CompletableFuture<Void> newStatus = status
+                // Unblock and log errors when task completes
+                .handle((result, error) -> {
+                    unblock();
 
-            // Log any errors
-            if(error != null){
-                LOG.error("Error while executing mutator", error);
-            }
+                    // Log any errors
+                    if (error != null) {
+                        printError("Error while executing mutator: \n" + error);
+                        if(debugOn) throw new RuntimeException(error);
+                    }
 
-            return result;
-        })
-        // Execute registered completion function
-        .thenAcceptAsync(onCompletionOfTask)
-        // Log errors in completion function
-        .exceptionally(t -> {
-            LOG.error("Error in callback for mutator", t);
-            throw new RuntimeException(t);
-        });
+                    return result;
+                })
+                // Execute registered completion function
+                .thenAcceptAsync(onCompletionOfTask);
 
+        // TODO: use an async client
+        futures.put(newStatus.hashCode(), newStatus);
     }
 
-    private void unblock(CompletableFuture<Json> status){
+    private void unblock(){
         blocker.release();
-        futures.remove(status.hashCode());
```

Comment:
We should remove the futures at _some_ point, or we'll have a leak. I guess I'd do it in `waitToFinish` after `get`?

Reply:
Ah yes good point.